### PR TITLE
Spec patch for the Web Audio API restriction

### PR DESCRIPTION
**Pull request by toyoshim**
This is based on the discussion at #165

### CODE REVIEW

File: prerendering.bs [hunk 1]
```diff
@@ -787,6 +794,20 @@ Add {{[DelayWhilePrerendering]}} to {{Navigator/getUserMedia()}}, {{MediaDevices
   1. If [=this=]'s [=relevant global object=]'s [=Window/navigable=] is a [=prerendering navigable=], then return.
 </div>
 
+<h4 id="web-audio-patch">Web Audio API</h4>
+
+The concept, [=allowed to start=], is used in the spec, but details are not defined as implementation details.
+To restrict auto playback while prerendering, add the following note and steps in the [=AudioContext=] interface section.
+
+<p class="note">
```
Comment (domenic) [inline prerendering.bs:802]: `class="note"` is for informative notes, not normative requirements. So just use a normal paragraph and say "the following requirements" instead.

File: prerendering.bs [hunk 2]
```diff
@@ -787,6 +794,20 @@ Add {{[DelayWhilePrerendering]}} to {{Navigator/getUserMedia()}}, {{MediaDevices
...
+The [=AudioContext=] is never allowed to start while prerendering.
```
Comment (domenic) [inline prerendering.bs:803]:
```suggestion
The [=AudioContext=] is never [=allowed to start=] while prerendering.
```

**Review by domenic — COMMENTED** [review]
Overall, this looks pretty reasonable to me. I left a few editorial comments.

I am a little scared of "If [=allowed to start=] algorithm runs and returns false only due to the prerendering". It is very general.

The alternative would be to only patch the AudioContext constructor. But, I am not sure if that is right either... the other three call sites (in `resume()` and `start(when)` and `start(when, offset, duration)`) might also be important. Do they need the same behavior, of setting the control thread state and queueing a control message to resume? Hmm.

So, I am OK with the text as-is, but would be interested in your thoughts. (Or thoughts from @hoch, if possible.)

**Comment by toyoshim** [comment]
`allow to start` needs to return `false` for all callers during prerendering, but the post-prerendering activation patch is probably needed only in the AudioContext constuctor case.

Users may or may not call other resume or start methods, but their promises would not be resolved until the post-prerendering activation steps run. When our step runs, the resume control message is processed, and it resolves all pending resume promises.

So, it seems to work without  compatibility breakages. I will revise my PR in this direction.

File: prerendering.bs (hunk 1 above)
Comment (toyoshim) [reply prerendering.bs:818]: done.

File: prerendering.bs (hunk 2 above)
Comment (toyoshim) [reply prerendering.bs]: done.

**Comment by hoch** [comment]
As you already saw, the construction of AudioContext, Autoplay control, and the resume behavior is rather messy.

I believe the last comment from @toyoshim looks correct.

File: prerendering.bs [hunk 3]
```diff
@@ -787,6 +802,30 @@ Add {{[DelayWhilePrerendering]}} to {{Navigator/getUserMedia()}}, {{MediaDevices
...
+<h4 id="web-audio-patch">Web Audio API</h4>
+
+The concept, [=allowed to start=], is used in the spec, but details are not defined as implementation details.
+To restrict auto playback while prerendering, add the following rule in the [=AudioContext=] interface section.
+
+The [=AudioContext=] is never [=allowed to start=] while prerendering.
+
+Also modify [=AudioContext constructor=] steps to have the following step before returning the constructed object.
+
+1. Else if the context is not [=allow to start=] only due to the prerendering, then append the following steps to [=this=]'s [=Document/post-prerendering activation steps list=].
```
Comment (hoch) [inline prerendering.bs:814]: What does it mean by "Else if" here?

File: prerendering.bs [hunk 4]
```diff
@@ -787,6 +795,30 @@ Add {{[DelayWhilePrerendering]}} to {{Navigator/getUserMedia()}}, {{MediaDevices
...
+  1. Set the [=[[control thread state]]=] on the {{AudioContext}} to [=running=].
+  1. [=Queue a control message=] to resume the {{AudioContext}}
+
+<p class="note">
+Users might call {{AudioContext/resume()}} that ensurs the context is not [=allow to start=] and appends a {{Promise}} to [=[[pending resume promises]]=].
+The {{Promise}} will be resolved in the activation steps above.
+</p>
+
+<p class="note">
+The {{AudioScheduledSourceNode}} interface has {{AudioScheduledSourceNode/start(when)}} method and the {{AudioBufferSourceNode}} interface has {{AudioBufferSourceNode/start(when, offset, duration)}} method that might affect [=[[control thread state]]=].
```
Comment (domenic) [inline prerendering.bs:817]:
```suggestion
The {{AudioScheduledSourceNode}} interface has a {{AudioScheduledSourceNode/start(when)}} method and the {{AudioBufferSourceNode}} interface has a {{AudioBufferSourceNode/start(when, offset, duration)}} method that might affect [=[[control thread state]]=].
```

File: prerendering.bs (hunk 3 above)
Comment (toyoshim) [reply prerendering.bs:807]: Yes.

**Comment by toyoshim** [comment]
Thanks!

File: prerendering.bs (hunk 3 above)
Comment (toyoshim) [reply prerendering.bs]: @hoch Are you ok with the current PR, or do you think we need some editing here to clarify the patch's context a little more?

Comment (hoch) [reply prerendering.bs]: The current PR looks good to me. Thanks for asking!

**Review by hoch — APPROVED** [review]
LGTM

**Pull request merged**